Updating Metrics for 1.3 Changes (#1954)
dbt-labs/docs.getdbt.com: website/docs/docs/building-a-dbt-project/metrics.md

<Changelog>

> * **v1.3.0**: Metrics are considered 

Reviewer: @callum-mcdata what does this mean? or are we trying to say Metrics are operational?

Author: :sweat: woopsies! I mean to say that they are no longer experimental

* **v1.0.0**: Metrics are new and experimental

</Changelog>

## About Metrics 

A metric is a timeseries aggregation over a <Term id="table" /> that supports zero or more dimensions. Some examples of metrics include:
- active users
- monthly recurring revenue (mrr)

> In v1.0, dbt supports metric definitions as a new node type. Like [exposures](exposures), metrics participate in the dbt docs lineage (DAG) and can be expressed in YAML files. By defining metrics in dbt projects, you encode crucial business logic in tested, version-controlled code. Further, you can expose these metrics definitions to downstream tooling, which drives consistency and precision in metric reporting.

Reviewer: spelled out dag for consistency purposes

```suggestion
In v1.0, dbt supports metric definitions as a new node type. Like [exposures](exposures), metrics appear as a directed acyclic graph (DAG) lineage in the documentation and can be expressed in YAML files. Defining metrics in dbt projects encodes crucial business logic in tested, version-controlled code. Further, you can expose these metrics definitions to downstream tooling, which drives consistency and precision in metric reporting.
```

v1.0.0 includes an initial version of metrics, following a [vibrant community discussion](https://github.com/dbt-labs/dbt-core/issues/4071). Try them out, and let us know what you think!
:::

## About Metrics 

A metric is a timeseries aggregation over a <Term id="table" /> that supports zero or more dimensions. Some examples of metrics include:
- active users
- monthly recurring revenue (mrr)

> In v1.0, dbt supports metric definitions as a new node type. Like [exposures](exposures), metrics participate in the dbt DAG and can be expressed in YAML files. By defining metrics in dbt projects, you encode crucial business logic in tested, version-controlled code. Further, you can expose these metrics definitions to downstream tooling, which drives consistency and precision in metric reporting.

Reviewer: ```suggestion
In v1.0, dbt supports metric definitions as a new node type. Like [exposures](exposures), metrics participate in the dbt docs lineage (DAG) and can be expressed in YAML files. By defining metrics in dbt projects, you encode crucial business logic in tested, version-controlled code. Further, you can expose these metrics definitions to downstream tooling, which drives consistency and precision in metric reporting.
```

    meta: {team: Finance}
```
</VersionBlock> 

</File>   

### Available properties

> Metrics can have a number of declared **properties**, which define aspects of your metric. More information on [properties and configs can be found here](https://docs.getdbt.com/reference/configs-and-properties).

Reviewer: ```suggestion
Metrics can have many declared **properties**, which define aspects of your metric. More information on [properties and configs can be found here](https://docs.getdbt.com/reference/configs-and-properties).
```

Author: sounds good to me

| sum            | This metric type will apply the `sum` aggregation to the specified field |
| average        | This metric type will apply the `average` aggregation to the specified field |
| min            | This metric type will apply the `min` aggregation to the specified field |
| max            | This metric type will apply the `max` aggregation to the specified field |
|<VersionBlock firstVersion="1.3">derived </VersionBlock> <VersionBlock lastVersion="1.2">expression </VersionBlock>  | <VersionBlock firstVersion="1.2"> This metric type is defined as any _non-aggregating_ calculation of 1 or more metrics </VersionBlock> |

<VersionBlock firstVersion="1.3">

### Derived Metrics
In v1.2, support was added for `derived` metrics (previously named `expression`), which are defined as non-aggregating calculations of 1 or more metrics. An example of this would be `{{metric('total_revenue')}} / {{metric('count_of_customers')}}`.

 By defining these metrics, you are able to create metrics like:
- ratios
- subtractions 
- any arbitrary calculation

> As long as the two+ base metrics (the metrics that comprise the `derived` metric) share the specified `time_grains` and `dimensions`, those attributes can be used in any downstream metrics macro.

Reviewer: ```suggestion
As long as the two or more base metrics (the metrics that comprise the `derived` metric) share the specified `time_grains` and `dimensions`, those attributes can be used in any downstream metrics macro.
```

    meta: {team: Finance}
```
</VersionBlock> 

</File>   

> ### Available properties

Reviewer: can we add contextual info on what properties are?

| Field       | Description                                                 | Example                         | Required? |
|-------------|-------------------------------------------------------------|---------------------------------|-----------|
| name        | A unique identifier for the metric                          | new_customers                   | yes       |
| model       | The dbt model that powers this metric                       | dim_customers                   | <VersionBlock firstVersion="1.3">yes (no for `derived` metrics)</VersionBlock><VersionBlock lastVersion="1.2">yes (no for `expression` metrics)</VersionBlock><VersionBlock lastVersion="1.1">yes</VersionBlock> |
| label       | A short for name / label for the metric                     | New Customers                   | no        |
| description | Long form, human-readable description for the metric        | The number of customers who.... | no        |

> | <VersionBlock firstVersion="1.3">calculation_method</VersionBlock><VersionBlock lastVersion="1.1">type </VersionBlock> | <VersionBlock firstVersion="1.3">The method of calculation (aggregation or derived) that is applied to the expression</VersionBlock><VersionBlock lastVersion="1.1">The type of calculation to perform when evaluating a metric</VersionBlock>  | count_distinct | yes       |

Reviewer: the table seems to be formatting oddly when switching versions:

**v1.2 rows under description are empty and the last row only says Not yet available - added in v1.3:**

![Screenshot 2022-09-09 at 10 07 57](https://user-images.githubusercontent.com/89008547/189315272-83ac99b2-94b9-4db2-90cf-5acfb243751c.png)

**v1.1 -> `models` row** has a trailing _yes:_  
![Screenshot 2022-09-09 at 10 11 04](https://user-images.githubusercontent.com/89008547/189315280-69256d94-de76-4f2d-89ec-79c8fa7c6453.png)

| timestamp   | The time-based component of the metric                      | signup_date                     | yes       |
| time_grains | One or more "grains" at which the metric can be evaluated   | [day, week, month]              | yes       |
| dimensions  | A list of dimensions to group or filter the metric by       | [plan, country]                 | no        |
| filters     | A list of filters to apply before calculating the metric    | See below                       | no        |
| meta        | Arbitrary key/value store                                   | {team: Finance}                 | no        |
|<VersionBlock firstVersion="1.3"> window </VersionBlock> <VersionBlock lastVersion="1.2">Not yet available - added in v1.3 </VersionBlock>     | <VersionBlock firstVersion="1.3"> Used for rolling/cumulative metrics where the calculation method is applied across a range of time  </VersionBlock> <VersionBlock lastVersion="1.2">Not yet available — added in v1.3</VersionBlock>  | <VersionBlock firstVersion="1.3"> 14 days  </VersionBlock> <VersionBlock lastVersion="1.2">Not yet available — added in v1.3</VersionBlock>   | no |

> ### Available types

Reviewer: can we add contextual info on what types are and what they do?

Author: These correspond to the type/calculation method property of the metric, where it is defined in the above table

| average        | This metric type will apply the `average` aggregation to the specified field |
| min            | This metric type will apply the `min` aggregation to the specified field |
| max            | This metric type will apply the `max` aggregation to the specified field |

> |<VersionBlock firstVersion="1.3">derived </VersionBlock> <VersionBlock lastVersion="1.2">expression </VersionBlock>  <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock>   | <VersionBlock firstVersion="1.2"> This metric type is defined as any **non-aggregating** calculation of 1 or more metrics </VersionBlock> <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock> |

Reviewer: ```suggestion
|<VersionBlock firstVersion="1.3">derived </VersionBlock> <VersionBlock lastVersion="1.2">expression </VersionBlock>  <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock>   | <VersionBlock firstVersion="1.2"> This metric type is defined as any _non-aggregating_ calculation of 1 or more metrics </VersionBlock> <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock> |
```

> |<VersionBlock firstVersion="1.3">derived </VersionBlock> <VersionBlock lastVersion="1.2">expression </VersionBlock>  <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock>   | <VersionBlock firstVersion="1.2"> This metric type is defined as any **non-aggregating** calculation of 1 or more metrics </VersionBlock> <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock> |

Reviewer: ```suggestion
|<VersionBlock firstVersion="1.3">derived </VersionBlock> <VersionBlock lastVersion="1.2">expression </VersionBlock>  <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock>   | <VersionBlock firstVersion="1.2"> This metric type is defined as any _non-aggregating_ calculation of 1 or more metrics </VersionBlock> <VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock> |
```

Author: I'm going to leave this as bold because I think its really important to call out that they are non aggregating. It has been a point of confusion for some people so want to make sure it is pretty in your face

<VersionBlock firstVersion="1.3">

### Derived Metrics

> In v1.2, support was added for `derived` (previously named expression) metrics, which are defined as non-aggregating calculations of 1 or more metrics. By defining these metrics, you are able to create metrics like:

Reviewer: to match style guide

```suggestion
In v1.2, support was added for `derived` metrics (previously named `expression`), which are defined as non-aggregating calculations of 1 or more metrics. An example of this would be `{{metric('total_revenue')}} / {{metric('count_of_customers')}}`.

 By defining these metrics, you are able to create metrics like:
```

        value: "'2020-01-01'"
```

## Querying Your Metric

> If you're interested in querying your metric, you must have the [dbt_metrics package](https://github.com/dbt-labs/dbt_metrics) installed. Information on how to install packages in [your project can be found here](https://docs.getdbt.com/docs/building-a-dbt-project/package-management#how-do-i-add-a-package-to-my-project) - this can be used in conjunction with the code snippet below to ensure you can query metrics in your project. 

Reviewer: ```suggestion
You can dynamically query metrics directly in dbt and verify them before running a job in the deployment environment.  To query your defined metric, you must have the [dbt_metrics package](https://github.com/dbt-labs/dbt_metrics) installed. 

Copy the below [metrics package](https://hub.getdbt.com/dbt-labs/metrics/latest/) code inside your dbt project in a packages.yml file and run `dbt deps` to install the metrics package:

```

Author: I will make a few edits here because the second sentence doesn't sound right for installing the package. Less using the code from the metrics package and more installation info.

Reviewer: gotcha good call, maybe it's just linking the package but making sure they know they have to run dbt deps?

<VersionBlock firstVersion="1.3" lastVersion="1.3">

```yml
packages:
  - package: dbt-labs/metrics
    version: [">=0.4.0", "<0.5.0"]
```

</VersionBlock>

<VersionBlock firstVersion="1.2" lastVersion="1.2">

```yml
packages:
  - package: dbt-labs/metrics
    version: [">=0.3.0", "<0.4.0"]
```

</VersionBlock>

<VersionBlock firstVersion="1.1" lastVersion="1.1">

```yml
packages:
  - package: dbt-labs/metrics
    version: [">=0.2.0", "<0.3.0"]
```

</VersionBlock>

### Querying metrics with `metrics.calculate`

> The calculate macro is used in conjunction with defined metrics to generate a sql statement that runs the metric aggregation to return the correct metric dataset. Example below:

Reviewer: changing to active voice

```suggestion
Use the `metrics.calculate` macro along with defined metrics to generate a SQL statement that runs the metric aggregation to return the correct metric dataset. Example below:
```

<VersionBlock firstVersion="1.2" >

```sql
select * 
from {{ metrics.calculate(
    metric('new_customers'),
    grain='week',
    dimensions=['plan', 'country']
) }}
``` 

</VersionBlock>

<VersionBlock lastVersion="1.1">

```sql
select * 
from {{ metrics.calculate(
    metric_name='new_customers',
    grain='week',
    dimensions=['plan', 'country']
) }}
``` 

</VersionBlock>

### Supported Inputs

> The example above does not display all of the potential inputs that can be provided to the macro. Certain pieces of functionality, such as secondary calculations, can be complicated to use - we recomemend navigating to the [package README](https://github.com/dbt-labs/dbt_metrics) for more in depth information around each of the inputs that is not covered in the table below.

Reviewer: fix typo and switch to active voice

```suggestion
The example above doesn't display all the potential inputs you can provide to the macro.

You may find some pieces of functionality, like secondary calculations, complicated to use. We recommend reviewing the [package README](https://github.com/dbt-labs/dbt_metrics) for more in-depth information about each of the inputs that are not covered in the table below
```

| grain       | `day`, `week`, `month` | The time grain that the metric will be aggregated to in the returned dataset | Required |
| dimensions  | [`plan`, `country`] | The dimensions you want the metric to be aggregated by in the returned dataset | Optional |
| start_date  | `2022-01-01` | Limits the date range of data used in the metric calculation by not querying data before this date | Optional |
| end_date    | `2022-12-31` | Limits the date range of data used in the metric claculation by not querying data after this date | Optional |
| where       | `plan='paying_customer'` | A sql statment, or series of sql statements, that alter the **final** CTE in the generated sql. Most often used to limit the data to specific values of dimensions provided | Optional |

### Developing metrics with `metrics.develop`

<VersionBlock firstVersion="1.3" >

> There may be times when you want to test what a metric might look like before defining it in your project. In these cases you should use the `develop` metric, which allows you to provide metric(s) in a contained yml in order to simulate what the metric might loook like if defined in your project.

Reviewer: typo and switch to active

```suggestion
There may be times you want to test what a metric might look like before defining it in your project. In these cases, use the `develop` metric, which allows you to provide metric(s) in a contained yml so you can simulate what a defined metric might look like in your project.
```

```sql
{% set my_metric_yml -%}
{% raw %}

metrics:
  - name: develop_metric
    model: ref('fact_orders')
    label: Total Discount ($)
    timestamp: order_date
    time_grains: [day, week, month]
    calculation_method: average
    expression: discount_total
    dimensions:
      - had_discount
      - order_country

{% endraw %}
{%- endset %}

select * 
from {{ metrics.develop(
        develop_yml=my_metric_yml,
        metric_list=['develop_metric']
        grain='month'
        )
    }}
```

> **Important Caveat** - The metric list input for this macro takes in the metric names themselves, not the `metric('name')` statement that the `calculate` macro uses.

Reviewer: switch to sentence case
```suggestion
**Important caveat** - The metric list input for the `metrics.develop` macro takes in the metric names themselves, not the `metric('name')` statement that the `calculate` macro uses. Using the example above:

✅ `develop_metric`
❌ `'new_customers'`

```

</VersionBlock>

<VersionBlock firstVersion="1.2" lastVersion="1.2" >

> There may be times when you want to test what a metric might look like before defining it in your project. In these cases you should use the `develop` metric, which allows you to provide a single metric in a contained yml in order to simulate what the metric might loook like if defined in your project.

Reviewer: typo and active voice
```suggestion
There may be times you want to test what a metric might look like before defining it in your project. In these cases, the `develop` metric, which allows you to provide a single metric in a contained yml so you can simulate what a defined metric might look like in your project.
```

```sql
{% set my_metric_yml -%}
{% raw %}

metrics:
  - name: develop_metric
    model: ref('fact_orders')
    label: Total Discount ($)
    timestamp: order_date
    time_grains: [day, week, month]
    type: average
    sql: discount_total
    dimensions:
      - had_discount
      - order_country

{% endraw %}
{%- endset %}

select * 
from {{ metrics.develop(
        develop_yml=my_metric_yml
        grain='month'
        )
    }}
```

</VersionBlock>

<VersionBlock lastVersion="1.1" >

> Functionality for `develop` is only supported in v1.2 and greater. Please navigate to those versions for information about this method of metric development.

Reviewer: ```suggestion
Functionality for `develop` is only supported in v1.2 and higher. Please navigate to those versions for information about this method of metric development.
```

        value: "'2020-01-01'"
```

## Querying Your Metric
You can dynamically query metrics directly in dbt and verify them before running a job in the deployment environment.  To query your defined metric, you must have the [dbt_metrics package](https://github.com/dbt-labs/dbt_metrics) installed. Information on how to [install packages can be found here](https://docs.getdbt.com/docs/building-a-dbt-project/package-management#how-do-i-add-a-package-to-my-project).

Use the following [metrics package](https://hub.getdbt.com/dbt-labs/metrics/latest/) installation code in your packages.yml file and run `dbt deps` to install the metrics package:

<VersionBlock firstVersion="1.3" lastVersion="1.3">

```yml
packages:
  - package: dbt-labs/metrics
    version: [">=0.4.0", "<0.5.0"]
```

</VersionBlock>

<VersionBlock firstVersion="1.2" lastVersion="1.2">

```yml
packages:
  - package: dbt-labs/metrics
    version: [">=0.3.0", "<0.4.0"]
```

</VersionBlock>

<VersionBlock firstVersion="1.1" lastVersion="1.1">

```yml
packages:
  - package: dbt-labs/metrics
    version: [">=0.2.0", "<0.3.0"]
```

</VersionBlock>

> Once the package has been installed with `dbt deps`, make sure to run the `dbt_metrics_calendar_model` model as this is required for macros used to query metrics. More information on this, and additional calendar functinality, can be found in the [project README](https://github.com/dbt-labs/dbt_metrics#calendar).

Reviewer: ```suggestion
Once the package has been installed with `dbt deps`, make sure to run the `dbt_metrics_calendar_model` model as this is required for macros used to query metrics. More information on this, and additional calendar functionality, can be found in the [project README](https://github.com/dbt-labs/dbt_metrics#calendar).
```

### Querying metrics with `metrics.calculate`
Use the `metrics.calculate` macro along with defined metrics to generate a SQL statement that runs the metric aggregation to return the correct metric dataset. Example below:

<VersionBlock firstVersion="1.2" >

```sql
select * 
from {{ metrics.calculate(
    metric('new_customers'),
    grain='week',
    dimensions=['plan', 'country']
) }}
``` 

</VersionBlock>

<VersionBlock lastVersion="1.1">

```sql
select * 
from {{ metrics.calculate(
    metric_name='new_customers',
    grain='week',
    dimensions=['plan', 'country']
) }}
``` 

</VersionBlock>

> ### Supported Inputs

Reviewer: ```suggestion
### Supported inputs
```
